src/main/java/org/genomicsdb/spark/GenomicsDBSchemaFactory.java

```diff
@@ -0,0 +1,208 @@
+    fields.add(DataTypes.createStructField("variantType", DataTypes.StringType, true));
+    fields.add(DataTypes.createStructField("refAllele", DataTypes.StringType, true));
+    fields.add(DataTypes.createStructField("alternateAlleles", 
+          DataTypes.createArrayType(DataTypes.StringType, true), true));
+    fields.add(DataTypes.createStructField("sampleNames", 
+          DataTypes.createArrayType(DataTypes.StringType, false), false));
+    fields.add(DataTypes.createStructField("GT", 
+          DataTypes.createArrayType(DataTypes.StringType, true), true));
+    return fields;
+  }
+ 
+  public static StructType defaultSchema(){
+    return DataTypes.createStructType(defaultFields());
+  }
+
+  public static List<StructField> extendSchema(StructField[] addFields){
+    List<StructField> defaults = defaultFields();
+    StructType additional = DataTypes.createStructType(addFields);
+    Set<String> names = new HashSet<String>(Arrays.asList(additional.fieldNames()));
+    List<StructField> fields = new ArrayList<>();
+    for (StructField f: defaults){
+      if (names.contains(f.name())){
+        continue;
+      }else{
+        fields.add(f);
+      }
+    }
+    fields.addAll(Arrays.asList(addFields));
+    return fields;
+  }
+
+  private List<StructField> extendSchemaWithVid(StructField[] addFields){
+    List<StructField> defaults = defaultFields();
+    List<StructField> fields = new ArrayList<>();
+    StructType additional = DataTypes.createStructType(fields);
+    Set<String> names =  new HashSet<String>(Arrays.asList(additional.fieldNames())); 
+    for (StructField f: defaults){
+      if (names.contains(f.name())){
+        continue;
+      }else{
+        fields.add(f);
+      }
+    }
+    fields.addAll(addFieldsWithVid(addFields));
+    return fields;
+  }
+
+  private List<StructField> addFieldsWithVid(StructField[] addFields){
+    List<StructField> fields = new ArrayList<>();
+    for (StructField f: addFields){
+      GenomicsDBVidSchema field = this.vidMap.get(f.name());
+      if (field == null) {
+        throw new RuntimeException("Schema field " + f.name() + " not available in vid");      
+      }
+      fields.add(DataTypes.createStructField(f.name(), DataType.fromDDL(field.getDataType()), f.nullable()));
+    }
+    return fields;
+  }
+
+  public static StructType buildSchema(StructField[] addFields){
+    return DataTypes.createStructType(extendSchema(addFields));
+  }
+
+  public StructType buildSchemaWithVid(StructField[] addFields){
+    return DataTypes.createStructType(extendSchemaWithVid(addFields));
+  }
+
+  private Map<String, GenomicsDBVidSchema> buildVidSchemaMap(String loader){
+    Map<String, GenomicsDBVidSchema> vMap = new HashMap<String, GenomicsDBVidSchema>();
+    String vidMapping = "";
+    try {
+      FileReader loaderReader = new FileReader(loader);
+      JSONParser parser = new JSONParser();
+      JSONObject objLoad = (JSONObject) parser.parse(loaderReader);
```

Comment:
Hmm - perhaps you can assume that the vid is always Protobuf compliant? Any objections @mlathara? Are there users who have vid JSON files that aren't Protobuf compliant?

Reply:
Unfortunately, we do have users with vids that aren't protobuf compliant. Perhaps we can look at some way to upgrade those vids to be compatible and then remove this code.

Reply:
This function was just moved from GenomicsDBDataSourceReader to here, so it matches what is in master now. It's not related to Spark3 changes, only to organizational changes. I think supporting only protobuf here could be considered as task for future improvement.